### PR TITLE
Support timeouts for futex

### PR DESCRIPTION
**Pull request by arihant2math**
The BITSET variant specifies a duration from the EPOCH, but the regular variant it just an ordinary duration.

### CODE REVIEW

File: usertest/src/main.rs [hunk 1]
```diff
@@ -174,44 +174,6 @@ fn test_write() {
     println!(" OK");
 }
 
-fn test_futex() {
```
Comment (hexagonal-sun) [inline usertest/src/main.rs:177]: Any reason we've removed these tests?  I quite like having primitive tests that call the syscall.  I have another PR cooking which adds some edge-case tests too for things like a bogus uaddr, ensuring the fault handling mechanism works correctly, etc.

**Review by hexagonal-sun — APPROVED** [review]
Looks good!  Just one comment about removal of primitive tests.  Once `test_parking_lot_mutex_timeout` is uncommented, happy to merge.

File: usertest/src/main.rs (hunk 1 above)
Comment (arihant2math) [reply usertest/src/main.rs:177]: Sure, I've added it back in.

**Comment by arihant2math** [comment]
Hmm, the parking lot test hangs, I don't think this is implemented quite right.

**Comment by arihant2math** [comment]
Seems like `parking_lot` is retrying even when the syscall returns an error due to timeout?

**Comment by hexagonal-sun** [comment]
My hunch is because of this:

```rust
fn test_clock_sleep() {
    const SLEEP_LEN: Duration = Duration::from_millis(100);

    print!("Testing clock and sleep syscalls ...");
    let now = Instant::now();
    sleep(SLEEP_LEN);
    dbg!(now.elapsed());
    assert!(now.elapsed() >= SLEEP_LEN);
    println!(" OK");
}
```

Even though this does actually sleep, but the duration returns:

```
[usertest/src/main.rs:20:5] now.elapsed() = 0ns

thread 'main' (4) panicked at usertest/src/main.rs:21:5:
assertion failed: now.elapsed() >= SLEEP_LEN
```

So parking_lot thinks it didn't sleep for any time at all,  therefore keeps trying until it has actually slept for 100ms.

**Comment by arihant2math** [comment]
~~I think `CLOCK_REALTIME` is also just reporting uptime instead of time from epoch?~~
I'll see if I can fix this one, perhaps sleeping is broken, since the date function seems to work.

**Comment by hexagonal-sun** [comment]
Okay, I have a fix:

```diff
diff --git a/src/clock/realtime.rs b/src/clock/realtime.rs
index 05f8a44..712210d 100644
--- a/src/clock/realtime.rs
+++ b/src/clock/realtime.rs
@@ -1,5 +1,5 @@
 use crate::{
-    drivers::timer::{Instant, now},
+    drivers::timer::{Instant, now, uptime},
     sync::SpinLock,
 };
 use core::time::Duration;
@@ -14,7 +14,7 @@ pub fn date() -> Duration {
         let duraton_since_ep_info = now - ep_info.1;
         ep_info.0 + duraton_since_ep_info
     } else {
-        Duration::new(0, 0)
+        uptime()
     }
 }
```

If the date epoch hasn't been set then we should fallback to uptime.  That fixes the test.

**Comment by arihant2math** [comment]
I'll see if I can query an approximate time from the RTC in a different PR.

**Comment by hexagonal-sun** [comment]
Yes, having proper RTC drivers would be good.

**Comment by arihant2math** [comment]
Applied.

**Comment by hexagonal-sun** [comment]
Perfect, if that fixes the tests on your machine, I'm happy to merge.

**Comment by arihant2math** [comment]
👍 It does.
